File: js/interpreter.js
// Interprete de instrucciones. Modulo: "./interpreter.js"

var M = require("./machine.js");
var instructions = require("./instructions.js");
instructions.load(this);

// program es una lista de instrucciones. Tiene la forma:
// [[Iload, 0, 1], [Iload, 1 2], [Iadd, 2, 0, 1]]

function exec (inst) {
  // gR: get register, sR: set register.
  function gR (i) { return M.state.regs[inst[i]]; }
  function sR (i, val) { M.state.regs[inst[i]] = val; }

  var advance = true;
  var docall = false;
  function setpc (i) {
    M.state.pc = inst[i];
    advance = false;
  }
  function jumpto (name) {
    var lbl = M.state.labels[name];
    //console.log("jumping to " + name + ": " + lbl);
    if (lbl === null || lbl === undefined) {
      throw new Error("Undefined label: " + name);
    }
    M.state.pc = lbl;
    advance = false;
  }
  switch (inst[0]) {
    case "mov": sR(1, gR(2)); break;
    case "add": sR(1, gR(2) + gR(3)); break;
    case "sub": sR(1, gR(2) - gR(3)); break;
    case "mul": sR(1, gR(2) * gR(3)); break;
    case "div": sR(1, gR(2) / gR(3)); break;
    case "mod": sR(1, gR(2) % gR(3)); break;
    case "inc": sR(1, gR(1) + 1); break;
    case "dec": sR(1, gR(1) - 1); break;

    case "setarg": M.args[inst[2]]=gR(1); break;
    case "getarg": sR(1,M.args[inst[2]]); break;

    case "eq": sR(1, gR(2) == gR(3)); break;
    case "lt": sR(1, gR(2) < gR(3)); break;
    case "lteq": sR(1, gR(2) <= gR(3)); break;
    case "gtz": sR(1, gR(2) > 0); break;

    //case "loadi": sR(1, inst[2]); break;
    case "call": docall = true; break;

    case "jump": jumpto(inst[1]); break;
    case "jumpif": if (gR(2)) { jumpto(inst[1]) }; break;
    case "jumpifn": if (!gR(2)) { jumpto(inst[1]) }; break;

    //case "loadstr": sR(1, inst[2]); break;
    case "print": console.log(gR(1)); break;
    case "module": sR(1, M.modules[inst[2]][inst[3]]); break;
    case "label": break;
    case "end": M.endFunc(); advance=false; break;
    case "loadval": sR(1, inst[2]); break; // Por ahora este en vez de loadi...
    default: throw new Error("Unrecognized instruction: " + inst[0]);
  }
  if (advance) {
    M.state.pc++;
  }
  if (docall) {
    mcall(gR(1), inst[2])
  }
}

function compileLabels () {
  M.state.labels = {};
  for (var i = 0; i < M.state.code.length; i++) {
    var inst = M.state.code[i]
    if (inst[0] == "label") {
      M.state.labels[inst[1]] = i;
    }
  };
  //console.log(M.state.labels);
}

function mcall (fun, len) {
  if (typeof fun == "function") {
    fun();
  } else {
    M.loadFunc(fun);
    compileLabels();
  }
}

function execLoop () {
  var S = M.state;
  var inst = S.code[S.pc];
  if (M.debug > 0) { console.log("  " +instructions.toString(inst)); }
  exec(inst);
  if (S.pc >= S.code.length) {
    M.run = false;
  }
  if (M.run) {
    setTimeout(execLoop, M.debug);
  }
}

exports.exec = exec;
exports.execLoop = execLoop;
exports.compileLabels = compileLabels;

File: rubyvm/interpreter.rb
require_relative 'machine'

module Interpreter
  # Una pequeña clase que funciona como azúcar sintáctico para acceder
  # a los registros de la máquina.
  class Reg
    attr_reader :jumped, :debuged
    def initialize inst
      @inst = inst
      @jumped = false
      @debuged = false
    end
    def [] index
      Machine.state.regs[@inst[index]]
    end
    def []= index, value
      Machine.state.regs[@inst[index]] = value
    end
    def jump index
      Machine.state.pc = Machine.state.labels[@inst[index]]
      @jumped = true
    end
    def debug str
      str = str.clone
      str.gsub! "$0", self[0].to_s
      str.gsub! "$1", self[1].to_s
      str.gsub! "$2", self[2].to_s
      str.gsub! "%0", @inst[0].to_s
      str.gsub! "%1", @inst[1].to_s
      str.gsub! "%2", @inst[2].to_s
      Machine.debug str
      @debuged = true
    end
  end

  # Método principal
  def self.exec inst
    r = Reg.new inst
    advance = true
    fun = nil

    case inst.name.to_sym
    when :mov; r[0] = r[1]; r.debug "%0 = $1"
    when :neg; r[0] = 0 - r[0]; r.debug "%0 = 0 - $1"
    when :inc; r[0] = r[0] + 1; r.debug "%0++"
    when :dec; r[0] = r[0] - 1; r.debug "%0--"
    when :add; r[0] = r[1] + r[2]; r.debug "%0 = $1 + $2"
    when :sub; r[0] = r[1] - r[2]; r.debug "%0 = $1 - $2"
    when :mul; r[0] = r[1] * r[2]; r.debug "%0 = $1 * $2"
    when :div; r[0] = r[1] / r[2]; r.debug "%0 = $1 / $2"
    when :gtz; r[0] = r[1] > 0   ; r.debug "%0 = $1 > 0"
    when :eq;  r[0] = r[1] == r[2]; r.debug "%0 = $1 == $2"
    when :neq; r[0] = r[1] != r[2]; r.debug "%0 = $1 != $2"
    when :lt;  r[0] = r[1] <  r[2]; r.debug "%0 = $1 < $2"
    when :lte; r[0] = r[1] <= r[2]; r.debug "%0 = $1 <= $2"

    when :setarg; Machine.args[inst[1]] = r[0]; r.debug "A-%1 = $0"
    when :getarg; r[0] = Machine.args[inst[1]]; r.debug "%0 = A-%1"

    when :loadval; r[0] = inst[1]; r.debug "%0 = %1"
    when :print; puts r[0]
    when :module; r[0] = Machine.modules[inst[1]][inst[2]]; r.debug "%0 = %1.%2"
    when :call; fun = r[0]; r.debug "CALL %0"
    when :label; # No hacer nada, pero la instrucción es válida.
    when :jump; r.jump 0; r.debug "JUMP %0"
    when :jumpif; r.jump 0 if r[1]; r.debug "JUMP %0: $1"
    when :jumpifn; r.jump 0 if not r[1]; r.debug "JUMP %0: not $1"
    when :end
      end_fun = true
      advance=false
    else; raise "Unrecognized instruction #{inst}"
    end

    if Machine.debug? and r.debuged
      sleep Machine.debug/100.0
    end

    # El orden de esto importa
    Machine.state.pc += 1 if advance and not r.jumped
    Machine.load_function fun if fun
    Machine.end_function if end_fun
  end

  def self.execute
    main = Machine.modules[:Main][:main]
    Machine.load_function main

    while Machine.run?
      self.exec Machine.state.inst
    end
  end
end